src/js/pouch-express.js

```diff
@@ -0,0 +1,288 @@
+ *   2. dbName: { data: ["path1", "path2"] } // Multiple files to be loaded, long notation, no custom options.
+ *   3. dbName: { data: ["path1", "path2"], dbOptions: { db: memdown } } // Long notation, including additional custom database options.
+ *   4. dbName: { dbOptions: { db: memdown} } // Long notation, no data, but with custom database options.
+ *   5. dbName: {} // No data, no custom options.
+ *
+ * @param dbDef - The definition of a single database in any of the above formats.
+ * @returns {Object} - An expanded record in Object form.
+ *
+ */
+gpii.pouch.express.expandDbDef = function (dbDef) {
+    var expandedDef = {};
+    if (typeof dbDef === "object" && dbDef !== null && dbDef !== undefined) {
+        expandedDef = dbDef;
+        if (expandedDef.data) {
+            expandedDef.data = gpii.pouch.node.expandPath(expandedDef.data);
+        }
+    }
+
+    return expandedDef;
+};
+
+
+/**
+ *
+ * Initialize our instance of express-pouchdb.
+ *
+ * @param that
+ * @returns {Object} - The expressPouchDB middleware.
+ *
+ */
+gpii.pouch.express.initExpressPouchdb = function (that) {
+    fluid.log("express pouchdb instance '", that.id, "' initalizing...");
+
+    // Create our base directory if it doesn't already exist.
+    if (that.options.baseDir && !fs.existsSync(that.options.baseDir)) {
+        fluid.log("Creating directory '", that.options.baseDir, "' for express pouchdb instance '", that.id, "'...");
+        fs.mkdirSync(that.options.baseDir);
+    }
+
+    // There are unfortunately options that can only be configured via a configuration file.
+    //
+    // To allow ourselves (and users configuring and extending this grade) to control these options, we create the file
+    // with the contents of options.pouchConfig before configuring and starting express-pouchdb.
+    fs.writeFileSync(that.options.expressPouchConfigPath, JSON.stringify(that.options.expressPouchConfig, null, 2));
+
+    that.PouchDB = PouchDB.defaults(that.options.dbOptions);
+    that.expressPouchdb = expressPouchdb(that.PouchDB, { configPath: that.options.expressPouchConfigPath });
+
+    return that.expressPouchdb;
+};
+
+/**
+ *
+ * Initialize all of the configured databases in `that.options.database`.
+ *
+ * @param that
+ * @returns {Promise} - A `fluid.promise.sequence` that will be resolved when all databases are initialized.
+ *
+ */
+gpii.pouch.express.initDbs = function (that) {
+    // We create our components programatically because we need:
+    //
+    //   1. To be notified when all databases are ready for use.  We do this with a `fluid.promise.sequence`.
+    //   2. A way to clean up each database later on.  We do this by retaining a list of database instances.
+    //   3. A way to know when all of them have been destroyed.  We do this with another `sequence` in our cleanup invoker (see below).
+    //
+    // I can think of ways to accomplish #2 with dynamic components, but not 1 and 3.
```

Comment:
The use of dynamicComponents would be more obviously declarative. This can be done via the "array" style of dynamicComponents sourced from that.options.databases. http://docs.fluidproject.org/infusion/development/SubcomponentDeclaration.html#dynamic-subcomponents-with-a-source-array You can then pick them up later using fluid.queryIoCSelector http://docs.fluidproject.org/infusion/development/IoCAPI.html#fluid-queryiocselector-root-selector-flat- and start the asynchronous operations after onCreate etc and